libs/SmartStore/src/com/salesforce/androidsdk/smartstore/store/DBOpenHelper.java

```diff
@@ -191,9 +198,12 @@ protected DBOpenHelper(String encryptionKey, Context context, String dbName) {
 		this.dbName = dbName;
 	}
 
-	protected static void loadLibsIfNeeded(Context context) {
-		if (libsLoaded.compareAndSet(false, true)) {
-			 System.loadLibrary("sqlcipher");
+	protected static void loadNativeLibrariesIfNeeded() {
```

Comment:
I can see how it avoids concurrent calls but how does it avoids loading the library more than once.

Should you do something like:
```java
isSqlCipherLibraryLoaded.lock();
try {
  if (!libsLoaded.compareAndSet(false, true)) {
      System.loadLibrary("sqlcipher");
  }   
} finally {
  isSqlCipherLibraryLoaded.unlock();
}
```		

Also wouldn't it be simpler to do
```java
synchronized(this.class) {
   ...
}
```java

Reply:
According to my research and the AI summary on Google, `loadLibrary` has built in protections against duplication.  So, our goal here becomes making sure we don't let dependent threads through while one is still running `loadLibrary`, which was the case before.

I also profiled `loadLibrary` and it runs hundreds of times faster on subsequent calls for the same library, so there's no performance impact and that also hints that it is a fast-return on those additional calls.  It runs in less than 10 ms., so we can enjoy reducing our code footprint for readability knowing this.  We can also move away from using both `synchronized` and the atomic values.

Good question.  I believe we've got that covered.

Reply:
Cool.

Don't you think the following would be more readable:
```java
synchronized(DBOpenHelper.class) { 
  System.loadLibrary("sqlcipher");
}
```

Reply:
Thanks!  Synchronizing on a monitor object or class would work and provide similar readability.  `Lock` provides some benefits beyond creating a critical section just with the synchronization keyword.

https://developer.android.com/reference/kotlin/java/util/concurrent/locks/Lock

One reason I read to at least create a dedicated monitor topic for each concurrency subject is so that different topics don't compete for the overall class/instance monitor.  That can pose future refactoring needs if there is more than one concurrency topic per instance/class.  A `Lock` instance provides benefits beyond a simple monitor object, also, which also is really cool when we migrate this Java to Kotlin 🚀